### PR TITLE
Feat/Node: Block wrapper and Basic block parents

### LINKED ISSUE #870: Change Block into BasicBlock - nodejs
Blocked by https://github.com/iotaledger/iota-sdk/issues/881 and https://github.com/iotaledger/iota-sdk/issues/884

https://github.com/iotaledger/tips-draft/blob/tip46/tips/TIP-0046/tip-0046.md#basic-block

### LINKED ISSUE #875: Add Block wrapper - nodejs
Blocked by https://github.com/iotaledger/iota-sdk/issues/872 and https://github.com/iotaledger/iota-sdk/issues/870

https://github.com/iotaledger/tips-draft/blob/tip46/tips/TIP-0046/tip-0046.md#block-wrapper

### LINKED ISSUE #881: Add new parents field to Block - nodejs
https://github.com/iotaledger/tips-draft/blob/tip46/tips/TIP-0046/tip-0046.md#basic-block

### PR DESCRIPTION
**Pull request by kwek20**
Fixes #881
Fixes #875
Fixes [#870](https://github.com/iotaledger/iota-sdk/issues/870)

### CODE REVIEW

File: bindings/nodejs/lib/types/block/basic_block.ts [hunk 1]
```diff
@@ -1,22 +1,30 @@
...
-import type { HexEncodedString } from '../utils/hex-encoding';
 import { Payload, PayloadDiscriminator } from './payload';
 import { Type } from 'class-transformer';
+import { StrongParents, WeakParents, ShallowLikeParents } from './parents';
 
 /**
- * Block layout.
+ * Basic Block layout.
  */
 export class Block {
```
Comment (DaughterOfMars) [inline bindings/nodejs/lib/types/block/basic_block.ts]: What about ValidationBlock?

Comment (kwek20) [reply bindings/nodejs/lib/types/block/basic_block.ts]: Different issue

Comment (DaughterOfMars) [reply bindings/nodejs/lib/types/block/basic_block.ts]: Sure, but shouldn't this structure now be a `BasicBlock` at least?

File: bindings/nodejs/lib/types/block/parents/index.ts [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+    Max extends number,
+    A extends (T | undefined)[] = [],
+    O extends boolean = false,
+> = O extends false
+    ? Min extends A['length']
+        ? TupMinMax<T, Min, Max, A, true>
+        : TupMinMax<T, Min, Max, [...A, T], false>
+    : Max extends A['length']
+    ? A
+    : TupMinMax<T, Min, Max, [...A, T?], false>;
```
Comment (Thoralf-M) [inline bindings/nodejs/lib/types/block/parents/index.ts]: Is there no easier way? :carlos:

File: bindings/nodejs/lib/types/block/basic_block.ts (hunk 1 above)
Comment (kwek20) [reply bindings/nodejs/lib/types/block/basic_block.ts]: ye i suppose so, 1 sec

File: bindings/nodejs/lib/types/block/parents/index.ts (hunk 2 above)
Comment (DaughterOfMars) [reply bindings/nodejs/lib/types/block/parents/index.ts]: Is this ASCII art?

File: bindings/nodejs/lib/types/block/basic_block.ts (hunk 1 above)
Comment (Thoralf-M) [reply bindings/nodejs/lib/types/block/basic_block.ts]: But isn't this rather the BlockWrapper that's shared between both block types?

Comment (DaughterOfMars) [reply bindings/nodejs/lib/types/block/basic_block.ts]: Could be, or it could be two different types with duplicated fields. Doesn't really matter for the bindings really.

File: bindings/nodejs/lib/types/block/parents/index.ts (hunk 2 above)
Comment (kwek20) [reply bindings/nodejs/lib/types/block/parents/index.ts]: I couldnt find one no :D otherwise serialising would be annoying

File: bindings/nodejs/lib/types/block/core/block_wrapper.ts [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+abstract class BlockWrapper {
+    /**
+     * The type of block
+     */
+    type: BlockType;
+
+    /**
+     * Protocol version of the block.
+     */
+    protocolVersion: number;
```
Comment (Alexandcoats) [inline bindings/nodejs/lib/types/block/core/block_wrapper.ts:25]: Should this be `version`? We rename it in our params.

File: bindings/nodejs/lib/types/block/parents/index.ts (hunk 2 above)
Comment (Thoralf-M) [reply bindings/nodejs/lib/types/block/parents/index.ts]: How is this related to serialization?

Comment (kwek20) [reply bindings/nodejs/lib/types/block/parents/index.ts]: Making fields just array does not ensure length. making them classes would serialize them differently i think.
Anyways this is just the equivalent of `BoxedSlicePrefix<BlockId, BoundedU8<MIN, MAX>`, which i think we will be using more often around nodejs instead of manually verifying at constructor/setter.

I will change the name to that :)

Comment (Thoralf-M) [reply bindings/nodejs/lib/types/block/parents/index.ts]: > Making fields just array does not ensure length

Yeah but we should never send an invalid Block from the Rust side and the other way around it will be validated on the Rust side
We also don't validate all the other things

File: bindings/nodejs/lib/client/client.ts [hunk 4]
```diff
@@ -345,7 +346,7 @@ export class Client {
     /**
      * Post block as raw bytes, returns the block ID.
      */
-    async postBlockRaw(block: Block): Promise<BlockId> {
+    async postBlockRaw(block: BlockWrapper): Promise<BlockId> {
```
Comment (DaughterOfMars) [inline bindings/nodejs/lib/client/client.ts]: Can we keep the old name?

Comment (kwek20) [reply bindings/nodejs/lib/client/client.ts]: Yea we discussed it on slack a bit, lets discuss it on next sprint planning and leave for now

File: bindings/nodejs/lib/types/block/parents/index.ts (hunk 2 above)
Comment (kwek20) [reply bindings/nodejs/lib/types/block/parents/index.ts]: Alright ive removed it

File: bindings/nodejs/examples/client/10-mqtt.ts [hunk 5]
```diff
@@ -1,7 +1,13 @@
 // Copyright 2021-2023 IOTA Stiftung
 // SPDX-License-Identifier: Apache-2.0
 
-import { Block, Client, initLogger } from '@iota/sdk';
+import {
+    BasicBlock,
+    BlockWrapper,
```
Comment (Thoralf-M) [inline bindings/nodejs/examples/client/10-mqtt.ts]: defined but never used, same for plainToInstance

File: bindings/nodejs/lib/types/block/core/block_wrapper.ts [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+        this.slotCommitmentId = slotCommitmentId;
+        this.latestFinalizedSlot = latestFinalizedSlot;
+        this.issuerId = issuerId;
+        this.signature = signature;
+    }
+
+    /**
+     * The type of block.
+     */
+    getType(): BlockType {
```
Comment (Thoralf-M) [inline bindings/nodejs/lib/types/block/core/block_wrapper.ts]: Do we want all the getters? I thought the fields are accessible directly already

Comment (kwek20) [reply bindings/nodejs/lib/types/block/core/block_wrapper.ts]: yea i wasnt sure either, but we seem to do it in the current code

Comment (Thoralf-M) [reply bindings/nodejs/lib/types/block/core/block_wrapper.ts]: I think in the current code in some places we didn't remove the getters because it would be a breaking change
Generally not sure what we want :smile:

Comment (abdulmth) [reply bindings/nodejs/lib/types/block/core/block_wrapper.ts]: I think we can remove the getters and keep properties public, if people modify the properties, it's on them  😅

Comment (abdulmth) [reply bindings/nodejs/lib/types/block/core/block_wrapper.ts]: We have it here https://github.com/iotaledger/iota-sdk/issues/966
we can also use `readonly` when necessary

File: bindings/nodejs/lib/types/block/core/block_wrapper.ts [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+     */
+    readonly networkId: bigint;
+    /**
+     * The time at which the block was issued. It is a Unix timestamp in nanoseconds.
+     */
+    readonly issuingTime: bigint;
+    /**
+     * The identifier of the slot to which this block commits.
+     */
+    readonly slotCommitmentId: SlotCommitmentId;
```
Comment (abdulmth) [inline bindings/nodejs/lib/types/block/core/block_wrapper.ts]: `SlotCommitmentId` is introduced in https://github.com/iotaledger/iota-sdk/pull/1005 , we should use that once that PR is merged.

File: bindings/nodejs/lib/types/block/core/block_wrapper.ts [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+    /**
+     * The slot index of the latest finalized slot.
+     */
+    readonly latestFinalizedSlot: bigint;
```
Comment (abdulmth) [inline bindings/nodejs/lib/types/block/core/block_wrapper.ts]: this should be of type `SlotIndex`, also after https://github.com/iotaledger/iota-sdk/pull/1005 is merged.

File: bindings/nodejs/lib/types/block/core/block_wrapper.ts [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+    readonly type: BlockType;
+
+    /**
+     * Protocol version of the block.
+     */
+    readonly protocolVersion: number;
+    /**
+     * Network identifier.
+     */
+    readonly networkId: bigint;
```
Comment (abdulmth) [inline bindings/nodejs/lib/types/block/core/block_wrapper.ts]: use `u64` after https://github.com/iotaledger/iota-sdk/pull/1005

File: bindings/nodejs/lib/types/block/core/block_wrapper.ts [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+    /**
+     * The time at which the block was issued. It is a Unix timestamp in nanoseconds.
+     */
+    readonly issuingTime: bigint;
```
Comment (abdulmth) [inline bindings/nodejs/lib/types/block/core/block_wrapper.ts]: same here

**Review by DaughterOfMars — DISMISSED** [review]

**Review by abdulmth — DISMISSED** [review]

**Review by DaughterOfMars — DISMISSED** [review]

**Review by abdulmth — DISMISSED** [review]

**Review by DaughterOfMars — APPROVED** [review]

File: bindings/nodejs/lib/types/block/core/block.ts [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+    /// A Verification block.
+    Verification = 1,
+}
+
+abstract class Block {
+    /**
+     * The type of block
+     */
+    readonly type: BlockType;
+
```
Comment (Thoralf-M) [inline bindings/nodejs/lib/types/block/core/block.ts:15]:
```suggestion
```

**Review by DaughterOfMars — DISMISSED** [review]

**Review by abdulmth — DISMISSED** [review]

**Review by DaughterOfMars — DISMISSED** [review]

File: bindings/nodejs/lib/types/block/core/block.ts [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+import { u64 } from '../../utils/type-aliases';
+
+/**
+ * All of the block types.
+ */
+enum BlockType {
+    /// A Basic block.
+    Basic = 0,
+    /// A Verification block.
+    Verification = 1,
```
Comment (Thoralf-M) [inline bindings/nodejs/lib/types/block/core/block.ts]:
```suggestion
    /// A Validation block.
    Validation = 1,
```

**Review by Thoralf-M — DISMISSED** [review]

**Review by abdulmth — DISMISSED** [review]

**Review by DaughterOfMars — DISMISSED** [review]

**Review by DaughterOfMars — DISMISSED** [review]

**Review by Thoralf-M — DISMISSED** [review]

File: bindings/nodejs/lib/types/block/core/basic-block.ts [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+
+import { Payload, PayloadDiscriminator } from '../payload';
+import { Type } from 'class-transformer';
+import { StrongParents, WeakParents, ShallowLikeParents } from '../parents';
+import { Block } from './block';
+
+/**
+ * Basic Block layout.
+ */
+export class BasicBlock extends Block {
```
Comment (thibault-martinez) [inline bindings/nodejs/lib/types/block/core/basic-block.ts]: I don't know about that. This kinda changes the definition of `BasicBlock`

Comment (thibault-martinez) [reply bindings/nodejs/lib/types/block/core/basic-block.ts]: Feel like this should be done through composability and not inheritance

Comment (DaughterOfMars) [reply bindings/nodejs/lib/types/block/core/basic-block.ts]: Looks like a basically identical structure to the block enum could be created using TS generics and unions.

Comment (kwek20) [reply bindings/nodejs/lib/types/block/core/basic-block.ts]: yea we discussed it, and the serialized layout would be different. but json serialization doesnt take order into account

Comment (DaughterOfMars) [reply bindings/nodejs/lib/types/block/core/basic-block.ts]: So are you saying it would work or not?

Comment (kwek20) [reply bindings/nodejs/lib/types/block/core/basic-block.ts]: it works like this :)

Comment (thibault-martinez) [reply bindings/nodejs/lib/types/block/core/basic-block.ts]: Not sure I understand what's being said here, but I don't really like the proposed version, it's not strictly implementing the TIP.

Comment (kwek20) [reply bindings/nodejs/lib/types/block/core/basic-block.ts]: Then please elaborate why not, i dont see whats missing

Comment (thibault-martinez) [reply bindings/nodejs/lib/types/block/core/basic-block.ts]: Nothing is missing, but a `BasicBlock` shouldn't extend a `Block` and have access to the base fields, that's not how the TIP is. Why can't we use composition instead of this ?

Comment (kwek20) [reply bindings/nodejs/lib/types/block/core/basic-block.ts]: Because then the serialized structure becomes different.

instead of 
```
BasicBlock {
   Block wrapper fields...
   Basic Block fields...
}
```

it would be
```
BasicBlock {
   BlockWrapper {
      fields...
   },
   Basic Block fields...
}
```

Besides the tip says:
`The two block types introduced in this TIP share some common fields which are contained in an outer wrapper. This Block Wrapper is only introduced as a way to deduplicate the definition of those fields. It is not a standalone block type.`

What im doing is exactly that. (Base Block is abstract thus cannot instantiate)

Comment (thibault-martinez) [reply bindings/nodejs/lib/types/block/core/basic-block.ts]: > it would be
> 
> ```
> BasicBlock {
>    BlockWrapper {
>       fields...
>    },
>    Basic Block fields...
> }
> ```

Surely this can be changed with a very simple custom JSON serialization function that removes the extra brackets, or another simple solution. I don't see this as a blocker.

Comment (DaughterOfMars) [reply bindings/nodejs/lib/types/block/core/basic-block.ts]: Yeah I don't know TS well enough to say, but I feel like the above isn't right anyway with composition. I'm pretty sure a union type like `BlockWrapper<BasicBlock> | BlockWrapper<ValidationBlock>` would work.

Comment (kwek20) [reply bindings/nodejs/lib/types/block/core/basic-block.ts]: Ive given it a try to flatten/unflatten a typed object, but its too much custom code. You guys want to give it a try? I dont think its worth the effort just because it extends a type and the object order is different

Comment (kwek20) [reply bindings/nodejs/lib/types/block/core/basic-block.ts]: rewrote it , we had a misunderstanding :=)

File: bindings/nodejs/examples/client/08-data-block.ts [hunk 14]
```diff
@@ -39,10 +40,10 @@ async function run() {
         const fetchedBlock = await client.getBlock(blockIdAndBlock[0]);
         console.log('Block data: ', fetchedBlock);
 
-        if (fetchedBlock instanceof BasicBlock) {
+        if (fetchedBlock.type === BlockType.Basic) {
             const basic = fetchedBlock as BasicBlock;
-            if (basic instanceof TaggedDataPayload) {
-                const payload = basic as TaggedDataPayload;
+            if (basic.block.payload instanceof TaggedDataPayload) {
```
Comment (DaughterOfMars) [inline bindings/nodejs/examples/client/08-data-block.ts]: maybe we should do the same here with the type field

**Review by DaughterOfMars — DISMISSED** [review]

File: bindings/nodejs/examples/client/08-data-block.ts (hunk 14 above)
Comment (kwek20) [reply bindings/nodejs/examples/client/08-data-block.ts]: Its tricky, since the BasicBlock is a type, i cannot do `instanceof` on it (types are erased at runtime)
But maybe it would be good to align. just dont know what is the standard

Comment (DaughterOfMars) [reply bindings/nodejs/examples/client/08-data-block.ts]: instanceof seems like it should be bad practice to me

File: bindings/nodejs/lib/types/block/parents/index.ts [hunk 15]
```diff
@@ -0,0 +1,19 @@
...
+/**
+ *  Array of Strongly referenced parents in the range of 1..8
+ */
+export type StrongParents = Parents;
+/**
+ *  Array of Weakly referenced parents in the range of 0..8
+ */
+export type WeakParents = Parents;
+/**
+ *  Array of Shallowly referenced parents in the range of 0..8
```
Comment (Thoralf-M) [inline bindings/nodejs/lib/types/block/parents/index.ts]: For BasicBlocks this is true, for Validation the max is higher

**Review by DaughterOfMars — DISMISSED** [review]

**Review by Thoralf-M — DISMISSED** [review]

File: bindings/nodejs/lib/types/block/core/index.ts [hunk 16]
```diff
@@ -0,0 +1,28 @@
...
+// SPDX-License-Identifier: Apache-2.0
+
+import { plainToInstance } from 'class-transformer';
+import { BasicBlock, BasicBlockData } from './basic-block';
+import { BlockType, BlockWrapper } from './block';
+
+export * from './block';
+export * from './basic-block';
+
+export type Block = BasicBlock;
```
Comment (abdulmth) [inline bindings/nodejs/lib/types/block/core/index.ts]: why is this needed?

File: bindings/nodejs/lib/types/block/core/basic-block.ts [hunk 17]
```diff
@@ -0,0 +1,39 @@
...
+import { Type } from 'class-transformer';
+import { StrongParents, WeakParents, ShallowLikeParents } from '../parents';
+import { BlockWrapper } from './block';
+
+export type BasicBlock = BlockWrapper<BasicBlockData>;
+
+/**
+ * Basic Block layout.
+ */
+export class BasicBlockData {
```
Comment (abdulmth) [inline bindings/nodejs/lib/types/block/core/basic-block.ts]: don't we need constructor for this?

File: bindings/nodejs/lib/types/block/core/basic-block.ts [hunk 18]
```diff
@@ -0,0 +1,39 @@
...
+    /**
+     * Blocks that are strongly directly approved, in the range of 1..8.
+     */
+    readonly strongParents!: StrongParents;
```
Comment (abdulmth) [inline bindings/nodejs/lib/types/block/core/basic-block.ts]: why the bang is needed?

**Review by abdulmth — COMMENTED** [review]
Some annoying questions :D

File: bindings/nodejs/lib/types/block/core/index.ts (hunk 16 above)
Comment (kwek20) [reply bindings/nodejs/lib/types/block/core/index.ts]: its for later when theres more block types, itll be BasicBlock | ValidationBlock

File: bindings/nodejs/lib/types/block/core/basic-block.ts (hunk 17 above)
Comment (kwek20) [reply bindings/nodejs/lib/types/block/core/basic-block.ts]: no, itll be deserialized, and building is done using the builder pattern

File: bindings/nodejs/lib/types/block/core/basic-block.ts (hunk 18 above)
Comment (kwek20) [reply bindings/nodejs/lib/types/block/core/basic-block.ts]: indicates itll always have a value and never undefined

File: bindings/nodejs/examples/client/08-data-block.ts (hunk 14 above)
Comment (kwek20) [reply bindings/nodejs/examples/client/08-data-block.ts]: im always scared of == in js, as they also have the === :D instanceof is the safe choice haha

Comment (DaughterOfMars) [reply bindings/nodejs/examples/client/08-data-block.ts]: It's not the safe choice if we're using aliases

**Review by abdulmth — DISMISSED** [review]

**Review by DaughterOfMars — DISMISSED** [review]

**Review by abdulmth — DISMISSED** [review]

**Review by Thoralf-M — DISMISSED** [review]

File: bindings/nodejs/lib/types/block/core/basic.ts [hunk 19]
```diff
@@ -0,0 +1,37 @@
+// Copyright 2023 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+import { Payload, PayloadDiscriminator } from '../payload';
+import { Type } from 'class-transformer';
+import { StrongParents, WeakParents, ShallowLikeParents } from '../parents';
+import { Block, BlockType } from '.';
```
Comment (Thoralf-M) [inline bindings/nodejs/lib/types/block/core/basic.ts]: 'BlockType' is defined but never used

Comment (Thoralf-M) [inline bindings/nodejs/lib/types/block/core/basic.ts]: 'BlockType' is defined but never used

File: bindings/nodejs/examples/client/08-data-block.ts [hunk 20]
```diff
@@ -38,9 +38,12 @@ async function run() {
         const fetchedBlock = await client.getBlock(blockIdAndBlock[0]);
         console.log('Block data: ', fetchedBlock);
 
-        if (fetchedBlock.payload instanceof TaggedDataPayload) {
-            const payload = fetchedBlock.payload as TaggedDataPayload;
-            console.log('Decoded data:', hexToUtf8(payload.data));
+        if (fetchedBlock.isBasic()) {
```
Comment (DaughterOfMars) [inline bindings/nodejs/examples/client/08-data-block.ts:41]: Is that you, Rust?

File: bindings/nodejs/lib/types/block/id.ts [hunk 21]
```diff
@@ -27,3 +27,5 @@ export type TokenId = HexEncodedString;
  * A Foundry ID represented as hex-encoded string.
  */
 export type FoundryId = HexEncodedString;
+
+export type IssuerId = HexEncodedString;
```
Comment (thibault-martinez) [inline bindings/nodejs/lib/types/block/id.ts:34]: Comment ?

File: bindings/nodejs/lib/types/block/context_input/context-input.ts [hunk 22]
```diff
@@ -11,17 +11,17 @@ enum ContextInputType {
      */
     COMMITMENT = 0,
     /**
-     * The context input type of a `BlockIssuanceCreditContextInput`.
+     * The context input kind of a `BlockIssuanceCreditContextInput`.
```
Comment (thibault-martinez) [inline bindings/nodejs/lib/types/block/context_input/context-input.ts]: We don't really have an issue in TS with type being a reserved keyword, not sure I'd change everything, even more since the type is called ContextInputType

File: bindings/nodejs/lib/types/block/parents/index.ts [hunk 23]
```diff
@@ -0,0 +1,19 @@
+// Copyright 2023 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+import { BlockId } from '../id';
+
+export type Parents = BlockId[];
+
+/**
+ *  Array of Strongly referenced parents.
```
Comment (thibault-martinez) [inline bindings/nodejs/lib/types/block/parents/index.ts]: I wouldn't capitalise Strongly and all, bit weird...

File: bindings/nodejs/lib/types/block/core/basic.ts [hunk 24]
```diff
@@ -0,0 +1,37 @@
...
+
+/**
+ * Basic Block layout.
+ */
+export class BasicBlock extends Block {
+    /**
+     * Blocks that are strongly directly approved, in the range of 1..8.
```
Comment (thibault-martinez) [inline bindings/nodejs/lib/types/block/core/basic.ts]: Where did you take these comments? REST TIP? These ranges sound like a headache to maintain if they change, I wouldn't mention them.

File: bindings/nodejs/examples/client/08-data-block.ts (hunk 20 above)
Comment (kwek20) [reply bindings/nodejs/examples/client/08-data-block.ts:41]: yea figured why the heck not make it easy to use ;)

File: bindings/nodejs/lib/types/block/context_input/context-input.ts (hunk 22 above)
Comment (kwek20) [reply bindings/nodejs/lib/types/block/context_input/context-input.ts]: accidental from rebase, fixed

File: bindings/nodejs/lib/types/block/core/basic.ts (hunk 24 above)
Comment (kwek20) [reply bindings/nodejs/lib/types/block/core/basic.ts]: Its like how we want to do the rust parents per block type. However as we dont do validation on nodejs i left it to the comment. This was taken from our rust code a few weeks back.
I think its good to know what to expect, and know that an empty array is fine. 
I could switch it to "this array can be empty" for weak and shallow, and "cannot be" for strong?

File: bindings/nodejs/lib/types/block/context_input/context-input.ts [hunk 25]
```diff
@@ -41,11 +41,11 @@ class CommitmentContextInput extends ContextInput {
 }
 
 /**
- * A Block Issuance Credit (BIC) Context Input provides the VM with context for the value of
+ * A Block Issuance Credit (BIC) Input provides the VM with context for the value of
```
Comment (thibault-martinez) [inline bindings/nodejs/lib/types/block/context_input/context-input.ts]: Put it back please

Comment (thibault-martinez) [inline bindings/nodejs/lib/types/block/context_input/context-input.ts]:
```suggestion
 * A Block Issuance Credit (BIC) Context Input provides the VM with context for the value of
```

**Review by thibault-martinez — CHANGES_REQUESTED** [review]
Wasm tests failing

File: bindings/nodejs/lib/types/block/context_input/context-input.ts [hunk 26]
```diff
@@ -21,7 +21,7 @@ enum ContextInputType {
 }
 
 abstract class ContextInput {
-    readonly type: ContextInputType;
+    type: ContextInputType;
```
Comment (thibault-martinez) [inline bindings/nodejs/lib/types/block/context_input/context-input.ts]: Why not readonly anymore ?

File: bindings/nodejs/lib/types/block/context_input/context-input.ts [hunk 27]
```diff
@@ -45,7 +45,7 @@ class CommitmentContextInput extends ContextInput {
  * the BIC vector of a specific slot.
  */
 class BlockIssuanceCreditContextInput extends ContextInput {
-    readonly accountId: AccountId;
+    accountId: AccountId;
```
Comment (thibault-martinez) [inline bindings/nodejs/lib/types/block/context_input/context-input.ts]: same

File: bindings/nodejs/lib/types/block/context_input/context-input.ts [hunk 28]
```diff
@@ -57,7 +57,7 @@ class BlockIssuanceCreditContextInput extends ContextInput {
  * A Reward Context Input indicates which transaction Input is claiming Mana rewards.
  */
 class RewardContextInput extends ContextInput {
-    readonly index: u16;
+    index: u16;
```
Comment (thibault-martinez) [inline bindings/nodejs/lib/types/block/context_input/context-input.ts]: same

File: bindings/nodejs/lib/types/block/core/basic.ts (hunk 24 above)
Comment (thibault-martinez) [reply bindings/nodejs/lib/types/block/core/basic.ts]: I don't know, even that sounds like too much information. People are supposed to use high level functions we provide so they don't care about validation rules.

Comment (kwek20) [reply bindings/nodejs/lib/types/block/core/basic.ts]: Well good thing "Blocks that are strongly directly approved" explains it all :trollface: ive removed the numbers

**Review by DaughterOfMars — DISMISSED** [review]

**Review by Thoralf-M — DISMISSED** [review]

File: bindings/nodejs/lib/types/block/core/basic.ts [hunk 29]
```diff
@@ -0,0 +1,37 @@
...
+     */
+    @Type(() => Payload, {
+        discriminator: PayloadDiscriminator,
+    })
+    readonly payload?: Payload;
+    /**
+     * The amount of mana the Account identified by IssuerID is at most
+     * willing to burn for this block.
+     */
+    readonly burnedMana!: string;
```
Comment (thibault-martinez) [inline bindings/nodejs/lib/types/block/core/basic.ts]: Why are we not using the u64 typedef?

File: bindings/nodejs/lib/types/block/core/index.ts [hunk 30]
```diff
@@ -0,0 +1,20 @@
...
+import { plainToInstance } from 'class-transformer';
+import { BasicBlock } from './basic';
+import { BlockType } from './block';
+import { BlockWrapper } from './wrapper';
+
+export * from './wrapper';
+export * from './basic';
+export * from './block';
+
+export const BlockDiscriminator = {
```
Comment (thibault-martinez) [inline bindings/nodejs/lib/types/block/core/index.ts:12]: Feels like it should be in the block.ts file

Comment (kwek20) [reply bindings/nodejs/lib/types/block/core/index.ts:12]: wil cause cyclic dep

**Review by Thoralf-M — DISMISSED** [review]

File: bindings/nodejs/lib/types/block/core/index.ts (hunk 30 above)
Comment (thibault-martinez) [reply bindings/nodejs/lib/types/block/core/index.ts:12]: If you feel like it, I would add a small comment about that then. I see us asking why over and over again.

File: bindings/nodejs/lib/types/block/core/wrapper.ts [hunk 31]
```diff
@@ -0,0 +1,95 @@
...
+    @Type(() => Block, {
+        discriminator: BlockDiscriminator,
+    })
+    readonly block!: Block;
+
+    /**
+     * The block signature; used to validate issuance capabilities.
+     */
+    @Type(() => Ed25519Signature)
+    readonly signature!: Ed25519Signature;
```
Comment (thibault-martinez) [inline bindings/nodejs/lib/types/block/core/wrapper.ts]: This should not be an ED25519

**Review by DaughterOfMars — DISMISSED** [review]

File: bindings/nodejs/lib/types/block/core/index.ts (hunk 30 above)
Comment (thibault-martinez) [reply bindings/nodejs/lib/types/block/core/index.ts:12]: Yes, no ?

**Review by Thoralf-M — DISMISSED** [review]

File: bindings/nodejs/lib/types/block/core/wrapper.ts [hunk 32]
```diff
@@ -0,0 +1,97 @@
...
+    }
+
+    /**
+     * Gets the block as an actual `BasicBlock`.
+     * NOTE: Will throw an error if the block is not a `BasicBlock`.
+     * @returns The block
+     */
+    asBasic(): BasicBlock {
+        if (this.isBasic()) {
+            return this as unknown as BasicBlock;
```
Comment (thibault-martinez) [inline bindings/nodejs/lib/types/block/core/wrapper.ts]: Is this correct? Shouldn't it return this.block as... ?

File: bindings/nodejs/lib/types/block/core/index.ts (hunk 30 above)
Comment (kwek20) [reply bindings/nodejs/lib/types/block/core/index.ts:12]: yes

File: bindings/nodejs/lib/types/block/core/wrapper.ts (hunk 32 above)
Comment (kwek20) [reply bindings/nodejs/lib/types/block/core/wrapper.ts]: 💯 nice find

**Review by DaughterOfMars — DISMISSED** [review]

**Review by DaughterOfMars — DISMISSED** [review]

File: bindings/nodejs/lib/types/block/core/index.ts [hunk 33]
```diff
@@ -0,0 +1,15 @@
...
+// SPDX-License-Identifier: Apache-2.0
+
+import { BlockType } from './block';
+import { BasicBlock } from './basic';
+
+export * from './block';
+export * from './basic';
+export * from './wrapper';
+
+// Here because in block.ts it causes a circular Dependecy
```
Comment (thibault-martinez) [inline bindings/nodejs/lib/types/block/core/index.ts]:
```suggestion
// Here because in block.ts it causes a circular dependency
```

**Review by thibault-martinez — APPROVED** [review]

**Review by Thoralf-M — APPROVED** [review]

File: bindings/nodejs/examples/client/08-data-block.ts (hunk 20 above)
Comment (Alex6323) [reply bindings/nodejs/examples/client/08-data-block.ts:41]: not rusty enough tho :joy: ? wen `match fetchedBlock` (or `if let`) :stuck_out_tongue_closed_eyes: ?

File: bindings/nodejs/lib/types/block/core/block.ts [hunk 34]
```diff
@@ -0,0 +1,43 @@
...
+    }
+
+    /**
+     * Gets the block as an actual `BasicBlock`.
+     * NOTE: Will throw an error if the block is not a `BasicBlock`.
+     * @returns The block
+     */
+    asBasic(): BasicBlock {
+        if (this.isBasic()) {
+            return this as unknown as BasicBlock;
```
Comment (Alex6323) [inline bindings/nodejs/lib/types/block/core/block.ts:38]: `return this as BasicBlock` doesn't work already?

**Review by Alex6323 — APPROVED** [review]

File: bindings/nodejs/lib/types/block/core/block.ts (hunk 34 above)
Comment (thibault-martinez) [reply bindings/nodejs/lib/types/block/core/block.ts:38]: No

Comment (thibault-martinez) [reply bindings/nodejs/lib/types/block/core/block.ts:38]: > Conversion of type 'this' to type 'BasicBlock' may be a mistake because neither type sufficiently overlaps with the other. If this was intentional, convert the expression to 'unknown' first.

**Review by abdulmth — APPROVED** [review]
